File: app/src/main/java/edu/gatech/spellastory/data/Stories.java
package edu.gatech.spellastory.data;

import android.content.res.AssetManager;
import android.text.TextUtils;

import org.apache.commons.lang3.NotImplementedException;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import edu.gatech.spellastory.domain.Category;
import edu.gatech.spellastory.domain.stories.Story;
import edu.gatech.spellastory.domain.stories.StoryBlank;
import edu.gatech.spellastory.domain.stories.StoryLine;
import edu.gatech.spellastory.domain.stories.StoryToken;

public class Stories {

    private Map<String, Story> storyMap;
    private Categories categoriesDb;

    public Stories(Categories categoriesDb, AssetManager assets, String... stories) throws IOException {
        this.categoriesDb = categoriesDb;
        storyMap = readStories(Arrays.asList(stories), assets);
    }

    private Map<String, Story> readStories(List<String> stories, AssetManager assets) throws IOException {
        Map<String, Story> map = new HashMap<>();

        for (String storyName : stories) {
            InputStreamReader reader = new InputStreamReader(assets.open("stories/" + storyName + ".txt"));
            BufferedReader br = new BufferedReader(reader);

            List<String> lines = new ArrayList<>();
            String line;
            while ((line = br.readLine()) != null) {
                lines.add(line);
            }
            String storyText = TextUtils.join("\n", lines);

            Story story = new Story();
            List<StoryToken> storyTokens = parseLine(storyText);
            story.addAll(storyTokens);

            map.put(storyName, story);
        }

        return map;
    }

    private List<StoryToken> parseLine(String line) {
        List<StoryToken> storyTokens = new ArrayList<>();
        Map<String, List<Category>> blankCategories = new HashMap<>();
        String[] stringTokens = line.split(" ");

        int i = 0;
        while (i < stringTokens.length) {
            String token = stringTokens[i];
            System.out.println(token);

            if (StoryLine.isAudio(token)) {
                i++;

                String audioFile = token.substring(1, token.length() - 1);

                List<String> lineText = new ArrayList<>();
                while (i < stringTokens.length && isText(stringTokens[i])) {
                    lineText.add(stringTokens[i++]);
                }
                String text = TextUtils.join(" ", lineText);

                storyTokens.add(new StoryLine(audioFile, text));
            } else if (StoryBlank.isBlank(token)) {
                token = token.substring(1, token.length() - 1);
                String[] blankTokens = token.split("-");
                String identifier = blankTokens[0];

                List<Category> categories;
                if (blankTokens.length > 1) {
                    String[] categoryStrings = blankTokens[1].split(",");
                    categories = categoriesDb.getCategories(categoryStrings);
                    blankCategories.put(identifier, categories);
                } else {
                    categories = blankCategories.get(identifier);
                }

                storyTokens.add(new StoryBlank(identifier, categories));
                i++;
            } else {
                throw new NotImplementedException("What kind of token is this? " + token);
            }
        }

        return storyTokens;
    }

    private static boolean isText(String string) {
        return !StoryLine.isAudio(string) && !StoryBlank.isBlank(string);
    }

    public Story getStory(String story) {
        return storyMap.get(story);
    }
}

File: app/src/main/java/edu/gatech/spellastory/data/Stories.kt
package edu.gatech.spellastory.data

import android.app.Activity
import android.content.Context
import android.util.Log
import edu.gatech.spellastory.domain.Category
import edu.gatech.spellastory.domain.stories.Story
import edu.gatech.spellastory.domain.stories.StoryBlank
import edu.gatech.spellastory.domain.stories.StoryLine
import edu.gatech.spellastory.domain.stories.StoryToken
import edu.gatech.spellastory.util.SingletonHolder
import org.apache.commons.lang3.NotImplementedException
import java.io.BufferedReader
import java.io.InputStreamReader

val Activity.StoriesDb
    get() = Stories.getInstance(this)

class Stories private constructor(context: Context) {

    val stories: List<Story> by lazy {
        val reader = BufferedReader(InputStreamReader(context.assets.open("coded_phonemes.csv")))
        val phonemes = read(context, "the_special_invention", "the_space_alien", "santas_mixed_up_helper_elf", "magical_wheels")
        reader.close()
        phonemes
    }

    fun getStory(title: String): Story? {
        return stories.find { it.title == title }
    }

    private fun read(context: Context, vararg stories: kotlin.String) = stories.mapNotNull { filename ->
        val reader = BufferedReader(InputStreamReader(context.assets.open("stories/$filename.txt")))
        val text = reader.lineSequence().joinToString(" \n")
        try {
            Story(filename, parseText(text, Categories.getInstance(context)))
        } catch (e: NotImplementedException) {
            Log.e("Stories", e.message + " in story " + filename)
            null
        }
//        Story(filename, parseHtml(text, Categories.getInstance(context)))
    }

//    private fun parseHtml(storyHtml: String, categoriesDb: Categories): List<StoryToken> {
//        val list = mutableListOf<StoryToken>()
//        val doc = Jsoup.parse(storyHtml)
//        doc.body().select("s-p").forEach { p ->
//            list.add(Newline())
//
//            var token: StoryToken? = null
//            p.children().forEach { a ->
//                if (a.tagName().equals("s-audio", true)) {
//                    if (token == null) token = StoryLine(a.attr("data-src"), a.text())
//                    else (token as StoryLine).text += a.text()
//                } else if (a.tagName().equals("s-blank", true)) {
//                    if (token == null) token = StoryBlank(a.className(), )
//                }
//            }
//        }
//
//        val text = Html.fromHtml(storyHtml, null, Html.TagHandler { opening, tag, output, xmlReader ->
//            if (tag.equals("s-p", true)) list.add(Newline())
//            else if (tag.equals("s-audio", true)) {
//
//            }
//        })
//    }

    @Throws(NotImplementedException::class)
    private fun parseText(storyText: String, categoriesDb: Categories): List<StoryToken> {
        val storyTokens = mutableListOf<StoryToken>()
        val tokens = storyText.split(" ")
        val blankCategories = mutableMapOf<kotlin.String, List<Category>>()

        var i = 0
        while (i < tokens.size) {
            val token = tokens[i]
            var valid = false

            isAudio(token)?.also { audioFile ->
                i++

                val text = mutableListOf<kotlin.String>()
                while (i < tokens.size && isText(tokens[i]))
                    text.add(tokens[i++])

                storyTokens.add(StoryLine(audioFile, text.joinToString(" ")))
                valid = true
            }

            isBlank(token)?.also { _token ->
                val blankTokens = _token.split("-")

                val identifier = blankTokens[0]

                val categories: List<Category>
                if (blankTokens.size > 1) {
                    val categoryStrings = blankTokens[1].split(",")
                    categories = categoriesDb.getCategories(*categoryStrings.toTypedArray())
                    blankCategories[identifier] = categories
                } else {
                    categories = blankCategories.getValue(identifier)
                }

                storyTokens.add(StoryBlank(identifier, categories))
                i++
                valid = true
            }

            if (isNewline(token)) {
                i++
                valid = true
            }

            if (!valid) throw NotImplementedException("Invalid token: $token")
        }

        return storyTokens
    }

    private fun isAudio(token: kotlin.String): kotlin.String? {
        val tempToken = cleanToken(token)
        if (tempToken.startsWith("{") && tempToken.endsWith("}")) return tempToken.drop(1).dropLast(1)
        return null
    }

    private fun isBlank(token: kotlin.String): kotlin.String? {
        val tempToken = cleanToken(token)
        if (tempToken.startsWith("[") && tempToken.endsWith("]")) return tempToken.drop(1).dropLast(1)
        return null
    }

    private fun cleanToken(string: kotlin.String): kotlin.String {
        var ret = string.trim()
        if (ret.isNotEmpty() && ret.last() in PUNCTUATION) ret = ret.dropLast(1)
        return ret
    }

    private fun isNewline(string: kotlin.String): Boolean = string == "\n"

    private fun isText(string: kotlin.String): Boolean =
            isAudio(string) == null && isBlank(string) == null

    companion object : SingletonHolder<Stories, Context>(::Stories) {
        private const val PUNCTUATION = ".,:;"
    }
}